backup/backup-disks.py

```diff
@@ -0,0 +1,145 @@
+backup_logger = logging.getLogger("backup")
+
+def list_disks(compute, project, zone):
+    """ Lists all persistent disks used by project """
+    all_disks = []
+    result = compute.disks().list(project=project, zone=zone).execute()
+    all_disks.extend(result['items'])
+
+    while 'nextPageToken' in result:
+        result = compute.disks().list(project=project, zone=zone, \
+            pageToken=result['nextPageToken']).execute()
+        all_disks.extend(result['items'])
+
+    return all_disks
+
+
+def list_snapshots(compute, project):
+    """ Lists all snapshots created for this project """
+    all_snapshots = []
+    result = compute.snapshots().list(project=project).execute()
+    all_snapshots.extend(result['items'])
+
+    while 'nextPageToken' in result:
+        result = compute.snapshots().list(project=project, \
+            pageToken=result['nextPageToken']).execute()
+        all_snapshots.extend(result['items'])
+
+    return all_snapshots
+
+
+def filter_disks_by_name(disks, name):
+    """ Takes in NAME, a predefined prefix to disk names, and filters 
+    disks to only returns those that have NAME """
+    filtered_disks = []
+    backup_logger.info("Filtering disks by prefix name: %s", name)
+    for disk in disks:
+        try:
+            if name in disk['name']:
+                filtered_disks.append(disk)
+        except KeyError:
+            sys.exit("Improperly formatted disks -- is your information correct?")
+    return filtered_disks
+
+
+def filter_disks_by_namespace(disks, namespace, namespace_dict_key):
+    """ Takes in NAMESPACE, a predefined string value, and filters
+    disks to only return those belong to NAMESPACE """
+    filtered_disks = []
+    backup_logger.info("Filtering disks belong to namespace: %s", namespace)
+    for disk in disks:
+        try:
+            disk_namespace = json.loads(disk['description'])[namespace_dict_key]
+            if disk_namespace == namespace:
+                filtered_disks.append(disk)
+        except (JsonError, KeyError):
+            continue
+    return filtered_disks
+
+
+def filter_snapshots_by_time(snapshots, retention_period):
+    """ Takes in RETENTION_PERIOD, a number of days, and filters
+    disks to return only those that are older than the retention_period
+    and should be deleted """
+    try:
+        backup_logger.info("Filtering snapshots that are older than %d days", retention_period)
+        old_snapshots = list(filter(lambda snapshot: \
+            __days_between_now_and_last_backup(snapshot['creationTimestamp'][:SNAPSHOT_DATESTRING_LEN]) > \
+                retention_period, snapshots))
+    except (TypeError, KeyError):
+        sys.exit("Attempted to filter invalid snapshots")
+    return old_snapshots
+
+
+def create_snapshot_of_disk(compute, disk_name, project, zone, body):
+    """ Creates a snapshot of the provided disk """
+    backup_logger.info("Creating snapshot for disk %s", disk_name)
```

Comment:
I am not sure of this, but does `.execute()` throw an exception when the request fails, or do you need to look into the response to determine that?

A (small) downside of backup script would be that backup fails without notice.